### DISCUSSION TITLE
What link should I use in my "Invite user" email template?

### DISCUSSION THREAD

**Discussion by lchoward** [supabase/supabase#21097] [Questions] [answered]
Hey everyone 👋🏻 ! I am hitting a wall here, hopefully missing something obvious.

I have an app with an "Invite teammate" function that I want to email the user a link to set their password. I'm getting tripped up on the link + set password part.

➕ **The "Invite teammate" button works and uses**
```
auth.admin.inviteUserByEmail(email, {
    data: metadata,
});
```

This ends up
- Creating an auth.users row (no password)
- Creating a public.users row
- Emailing the user with the "Invite user" template

➖ **The rest is not working as intended**

The link pattern I have is:
`href="https://xxx.supabase.co/auth/v1/verify?token={{ .TokenHash }}&type=invite&redirect_to=https://mysite.com/set-password"`

My `/pages/set-password.tsx` (Next.js app) uses the Supabase Auth UI's "update_password" view

When the invitation email's link is clicked, the user appears to be successfully verified and redirected to `/set-password`. However, when I set a password and click "Update password" I get an error "Auth session missing!"

❓ Any idea what I'm missing here?

**Comment by mwoss** [comment]
The user needs to be authenticated to change their password. Supabase provides a short-lived access token (`access_token`) as a query param along with a redirect URL. The client needs to consume that access token, authorize and then update the password.

**Reply by lchoward** [reply]
@mwoss That's helpful fundamental background. Specifically for me then how would you recommend adjusting the email link so that the client can consume the access token?

This is what I have currently in my email template `href="https://xxx.supabase.co/auth/v1/verify?token={{ .TokenHash }}&type=invite&redirect_to=https://mysite.com/set-password"`

I've also tried `href="{{ .ConfirmationURL }}/set-password"`, which seems to be the same and not work.

**Reply by mwoss** [reply]
The password reset link seems to be fine (personally, I've always used `{{ .ConfirmationURL }}` and specified the `redirectTo` param). `access_token` will be appended to the redirect URL, and you can then retrieve that information from it and use the `supabase.auth.setSession` to authenticate a user with this data.

**Reply by lchoward** [reply]
Ah ok, trying that now. For some reason the `access_token` is not being appended to the redirect URL 🤔

**Reply by GaryAustin1** [reply]
How are you determining that?  It will not show up after the client grabs it and it can't be passed to serverside code as it is behind a hash (#).
You may "just" be losing your session thru the auth-helpers/ssr process if you are doing that.

**Reply by lchoward** [reply, marked as answer]
I ended up ditching the [Supabase Auth UI "update password" view](https://supabase.com/docs/guides/auth/auth-helpers/auth-ui#supported-views). Instead I used a basic password form (single field), and the end-to-end flow worked 🥳  

**In sum, made four changes to get this working**
1. Specified the `redirectTo` param for `inviteUserByEmail`
2. Adjusted the email template href to `{{ .ConfirmationURL }}` — automatically takes care of the verification, redirect, and `access_token` ✨ 
3. For my /pages/set-password.tsx (Next.js), I used my own form instead of the [Supabase Auth UI "update password" view](https://supabase.com/docs/guides/auth/auth-helpers/auth-ui#supported-views)
4. Added a `/pages/api/update-password.ts` to handle the request from the frontend. I used `auth.updateUser({password: newPassword})` here

> It will not show up after the client grabs it and it can't be passed to serverside code as it is behind a hash (#).

Makes sense!

**Reply by GaryAustin1** [reply]
You did not mention auth-ui I don't think...EDIT I missed you did.... 
 A struggle with password reset.
https://github.com/supabase/auth-ui/issues/77

**Reply by lchoward** [reply]
Thank you for the clear and timely help @mwoss @GaryAustin1 !!! Very appreciative of you two 🙏🏻

**Comment by GaryAustin1** [comment]
You probably should be doing the redirectTo in your api invite call as an option and not messing with the default URL in the template.
![image](https://github.com/supabase/supabase/assets/54564956/dea64f00-3c2c-4bef-8f8f-31c3419a06b0)

Unless you have messed with template string urls with redirect before.

**Reply by lchoward** [reply]
Thanks for this recommendation! 

@mwoss commented something very similar too. I'm getting closer, but hitting [this issue](https://github.com/orgs/supabase/discussions/21097#discussioncomment-8400984).
